langstream-api-gateway/src/main/java/ai/langstream/apigateway/http/GatewayResource.java

```diff
@@ -83,19 +88,102 @@ ProduceResponse produce(
             throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, e.getMessage());
         }
 
-        final ProduceGateway produceGateway =
+        try (final ProduceGateway produceGateway =
                 new ProduceGateway(
                         topicConnectionsRuntimeRegistryProvider
-                                .getTopicConnectionsRuntimeRegistry());
-        try {
+                                .getTopicConnectionsRuntimeRegistry()); ) {
             final List<Header> commonHeaders =
                     ProduceGateway.getProducerCommonHeaders(
                             context.gateway().getProduceOptions(), authContext);
             produceGateway.start(context.gateway().getTopic(), commonHeaders, authContext);
             produceGateway.produceMessage(produceRequest);
             return ProduceResponse.OK;
-        } finally {
-            produceGateway.close();
         }
     }
+
+    private Map<String, String> computeHeaders(WebRequest request) {
+        final Map<String, String> headers = new HashMap<>();
+        request.getHeaderNames()
+                .forEachRemaining(name -> headers.put(name, request.getHeader(name)));
+        return headers;
+    }
+
+    @PostMapping(
+            value = "/service/{tenant}/{application}/{gateway}",
+            consumes = MediaType.APPLICATION_JSON_VALUE)
+    void service(
+            WebRequest request,
+            HttpServletResponse response,
+            @NotBlank @PathVariable("tenant") String tenant,
+            @NotBlank @PathVariable("application") String application,
+            @NotBlank @PathVariable("gateway") String gateway,
+            @RequestBody ProduceRequest produceRequest)
```

Comment:
where do you set the timeout ?
I guess that we should give the client the ability to set the timeout for waiting for the message to arrive